### PR TITLE
[add] radarr_list plugin

### PR DESCRIPTION
**Pull request by paranoidi**
### Motivation for changes:

(Not my PR, continued from PR #1897 by @h0od, I just made it pretty)

A list plugin similar to the already existing Sonarr plugin but for Radarr which targets movies instead of TV shows.

The plugin acts both as an input plugin and as a managed list plugin.

### Detailed changes:
- A simple test is included which tests the input plugin functionality and all the list actions (add, remove, clear, match). A VCR cassette for the HTTP communication during test is also provided, so no live Radarr server is required during testing. **Be careful if running the test against a live server because all the movies will be cleared by the list_clear action**
- When matching entries, we compare in this order (1) IMDB and TMDB id's, (2) movie_name+movie_year, (3) title
- If setting `include_data: yes` when reading the list, all the entries generated will be decorated with attribute `quality_req`, which holds a set of quality requirements. These are transformed from the quality profile of the movie in Radarr to Fleget's own quality requirements. The simple mapping can be found [here](https://github.com/h0od/Flexget/commit/b8dc31cf5a9c0bd8b5797df0156d8f9bc21abc22#diff-10c57dd300d3be6f86a888887b303271R15)
- In Radarr, a quality profile can have a 'cut-off' quality. I realised as Flexget doesn't have a quality target/timeframe functionality for movies that the user might want to limit the quality requirements to just the cutoff quality. So by setting `only_use_cutoff_quality: yes`, the ´quality_req´ attribute is only set to the cut-off quality of the profile, and not all quality requirements.

#### Up for discussion:

- When using `list_add` to add entries to Radarr, the current solution provides no way to specify what quality profile the movie should be added with. We just add it to the default quality profile. Do we need such possibility?
- Is there a better way to map the [quality requirements](https://github.com/h0od/Flexget/commit/b8dc31cf5a9c0bd8b5797df0156d8f9bc21abc22#diff-10c57dd300d3be6f86a888887b303271R15)? 
- Because the entires in Radarr has a quality requirement, shouldn't we consider the quality of the entry when matching the entries in the list?

### Implemented feature requests:
- Feathub #[8](http://feathub.com/Flexget/Flexget/+8).

### Config usage
```
        tasks:
          clear_and_add_to_radarr_list:
            list_clear:
              what:
                - radarr_list:
                    base_url:  http://...
                    api_key: ...
                    port: 7878
            mock:
              - { title: 'Despicable Me 2 (2013)', imdb_id: 'tt1690953', tmdb_id: 93456 }
              - { title: 'Sinister 2 (2015)', imdb_id: 'tt2752772', tmdb_id: 283445 }
              - { title: 'Crimson Peak (2015)', imdb_id: 'tt2554274', tmdb_id: 201085 }
              - { title: 'Deadpool (2016)', imdb_id: 'tt1431045', tmdb_id: 293660 }
            accept_all: yes
            list_add:
              - radarr_list:
                  base_url:  http://...
                  api_key: ...
                  port: 7878

          radarr_list_as_input_plugin:
            radarr_list:
              base_url:  http://...
              api_key: ...
              port: 7878
              include_data: True
            accept_all: yes

          remove_from_radarr_list:
            mock:
              - { title: "Ocean's Twelve (2004)", imdb_id: 'tt0349903', tmdb_id: 163 }
              - { title: 'Sinister 2 (2015)', imdb_id: 'tt2752772', tmdb_id: 283445 }
            accept_all: yes
            list_remove:
              - radarr_list:
                  base_url:  http://...
                  api_key: ...
                  port: 7878

          match_radarr_list:
            mock:
              - { title: 'Despicable.Me.2.2013.1080p.BluRay.x264-FlexGet', imdb_id: 'tt1690953', tmdb_id: 93456 }
              - { title: 'Sinister.2.2015.720p.BluRay.x264-FlexGet', imdb_id: 'tt2752772', tmdb_id: 283445 }
              - { title: 'Crimson.Peak.2015.720p.BluRay.x264-FlexGet', imdb_id: 'tt2554274', tmdb_id: 201085 }
              - { title: 'Deadpool.2016.1080p.BluRay.x264-FlexGet', imdb_id: 'tt1431045', tmdb_id: 293660 }
              - { title: 'Kung.Fu.Panda.3.2016.720p.BluRay.x264-FlexGet', imdb_id: 'tt2267968', tmdb_id: 140300 }
            list_match:
              from:
                - radarr_list:
                    base_url:  http://...
                    api_key: ...
                    port: 7878
````

### CODE REVIEW

**Comment by paranoidi** [comment]
FYI, @h0od your excellent work has been continued here

File: flexget/plugins/list/radarr_list.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+
+log = logging.getLogger('radarr')
+
+
+@python_2_unicode_compatible
+class RadarrRequestError(Exception):
+    def __init__(self, value, logger=log, **kwargs):
+        super(RadarrRequestError, self).__init__()
+        # Value is expected to be a string
+        if not isinstance(value, str):
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: If it's expected, then maybe just condense it to `self.value = str(value)`?

File: flexget/plugins/list/radarr_list.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+    if "error_message" in radarr_request_ex.kwargs:
+        error_message = radarr_request_ex.kwargs["error_message"]
+
+    if not error_message:
+        return None
+
+    if error_message.lower() == "this movie has already been added":
+        return RadarrMovieAlreadyExistsError()
+
+    return None
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: `return None` is the default if there's no return statement

File: flexget/plugins/list/radarr_list.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+
+
+def request_get_json(url, headers):
+    """ Makes a GET request and returns the JSON response """
+    try:
+        response = requests.get(url, headers=headers)
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: Shouldn't we always use `task.requests`?

File: flexget/plugins/list/radarr_list.py [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+                "Invalid response received from Radarr: %s" % response.content
+            )
+    except RequestException as e:
+        raise RadarrRequestError("Unable to connect to Radarr at %s. Error: %s" % (url, e))
+
+
+def request_delete_json(url, headers):
+    """ Makes a DELETE request and returns the JSON response """
+    try:
+        response = requests.delete(url, headers=headers)
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: See above

File: flexget/plugins/list/radarr_list.py [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+                "Invalid response received from Radarr: %s" % response.content
+            )
+    except RequestException as e:
+        raise RadarrRequestError("Unable to connect to Radarr at %s. Error: %s" % (url, e))
+
+
+def request_post_json(url, headers, data):
+    """ Makes a POST request and returns the JSON response """
+    try:
+        response = requests.post(url, headers=headers, data=data)
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: See above

File: flexget/plugins/list/radarr_list.py [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+            )
+
+    except RequestException as e:
+        raise RadarrRequestError("Unable to connect to Radarr at %s. Error: %s" % (url, e))
+
+
+def request_put_json(url, headers):
+    """ Makes a PUT request and returns the JSON response """
+    try:
+        response = requests.put(url, headers=headers)
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: See above.

It also seems all these requests methods basically do the same, so would it make sense to create a single function that takes in a request type as argument?

File: flexget/plugins/list/radarr_list.py [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+        """ Returns all movies that matches the search term """
+        term = quote(term)
+        request_url = self.api_url + "movies/lookup?term=" + term
+        headers = self._default_headers()
+        json_response = request_get_json(request_url, headers)
+        return json_response
+
+    def lookup_by_imdb(self, imdb_id):
+        """ Returns all movies that matches the imdb id """
+        # TODO: make regexp check that imdb_id really is an IMDB_ID
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: We have one in a util function somewhere.

File: flexget/plugins/list/radarr_list.py [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+        request_url = self.api_url + "movies/lookup/imdb?imdbId=" + imdb_id
+        headers = self._default_headers()
+        json_response = request_get_json(request_url, headers)
+        return json_response
+
+    def lookup_by_tmdb(self, tmdb_id):
+        """ Returns all movies that matches the tmdb id """
+        tmdb_id = int(tmdb_id)
+        request_url = self.api_url + "movies/lookup/tmdb?tmdbId=" + str(tmdb_id)
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: This is a little weird. int then str? Why? Remove leading zeros?

File: flexget/plugins/list/radarr_list.py [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+        Finds a movie by first checking agains the ids of the provided entry,
+        and if none matches, check by title name
+        """
+        for movie_entry in self.items:
+            # First check if any of the id attributes match
+            for id_attribute in ["tmdb_id", "imdb_id", "radarr_id"]:
+                if id_attribute in entry and id_attribute in movie_entry:
+                    if entry[id_attribute] == movie_entry[id_attribute]:
+                        # Perfect match!
+                        return movie_entry
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: A lot of indentation here.
```suggestion
            for id_attribute in ["tmdb_id", "imdb_id", "radarr_id"]:
                if id_attribute in entry and entry.get(id_attribute) and movie_entry.get(id_attribute):
                    # Perfect match!
                    return movie_entry```

File: flexget/plugins/list/radarr_list.py [hunk 10]
```diff
@@ -0,0 +1,593 @@
...
+
+            # Then we check if the title matches
+            movie_name = entry.get("movie_name")
+            movie_year = entry.get("movie_year")
+            if movie_name:
+                if movie_name.lower() == movie_entry["movie_name"].lower():
+                    # The name matches. If we also have a year lets check that as well.
+                    if movie_year and movie_entry["movie_year"]:
+                        if movie_year == movie_entry["movie_year"]:
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: These two checks seem a little redundant

File: flexget/plugins/list/radarr_list.py [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+                        # The movie had no year present
+                        return movie_entry
+
+            # Last resort is just to compare the title straight off
+            title = entry.get("title").lower()
+            if title == movie_entry["title"].lower():
+                return movie_entry
+
+        # Did not find a match
+        return None
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: Default

File: flexget/plugins/list/radarr_list.py [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+                        "The lookup for '%s' returned %d results. Using the first result '%s'.",
+                        title,
+                        len(results),
+                        results[0]["title"],
+                    )
+                    return results[0]
+            except RadarrRequestError as ex:
+                log.error("The search term lookup raised exception: %s", ex)
+
+        return None
```
Comment (cvium) [inline flexget/plugins/list/radarr_list.py]: Default

File: flexget/plugins/list/radarr_list.py (hunk 2 above)
Comment (paranoidi) [reply flexget/plugins/list/radarr_list.py]: Some say that having these is better for readability, I don't really have strong opinion in one way or another. Removed as we don't generally have them.

File: flexget/plugins/list/radarr_list.py (hunk 3 above)
Comment (paranoidi) [reply flexget/plugins/list/radarr_list.py]: Yeah, is it useful for application APIs though? Would need a bit refactoring ..

File: flexget/plugins/list/radarr_list.py (hunk 8 above)
Comment (paranoidi) [reply flexget/plugins/list/radarr_list.py]: Yeah, dunno. I just added str(..) around it as ide was complaining about it

File: flexget/plugins/list/radarr_list.py (hunk 10 above)
Comment (paranoidi) [reply flexget/plugins/list/radarr_list.py]: cleaned up

File: flexget/plugins/list/radarr_list.py (hunk 6 above)
Comment (paranoidi) [reply flexget/plugins/list/radarr_list.py]: Agreed, I did not really like these.

**Comment by paranoidi** [comment]
Merged, pushed  via other means

**Pull request closed without merging**